analyzers/ImpactParameter.py

```diff
@@ -1,48 +1,233 @@
+            #if TCHE > 30 and event.K0s == 0 and event.Kp == 0 and event.L0 == 0 and event.S0 == 0 and event.Sp == 0 and event.Sm == 0:
+            #    import pdb; pdb.set_trace()
+            
+            jet.tags['IP_b_LL'] = IP_b_LL  if self.tag_IP_b_LL  else None
+            jet.tags['IPs_b_LL']= IPs_b_LL if self.tag_IPs_b_LL else None
+            jet.tags['TCHE'] = TCHE
+            jet.tags['TCHP'] = TCHP
+            jet.tags['TCHE_IP'] = TCHE_IP
+            jet.tags['TCHP_IP'] = TCHP_IP
+            jet.tags['TCHE_x'] = TCHE_x
+            jet.tags['TCHE_y'] = TCHE_y
+            jet.tags['TCHE_z'] = TCHE_z
+            jet.tags['TCHE_xy'] = (TCHE_x**2+TCHE_y**2)**0.5
+            jet.tags['TCHE_pt'] = TCHE_pt
+            jet.tags['TCHE_dr'] = TCHE_dr
+            
+            if hasattr(event, 'K0s') == True :
+                jet.tags['K0s'] = event.K0s
+            else :
+                jet.tags['K0s'] = -99
+            if hasattr(event, 'Kp') == True :
+                jet.tags['Kp'] = event.Kp
+            else :
+                jet.tags['Kp'] = -99
+            if hasattr(event, 'L0') == True :
+                jet.tags['L0'] = event.L0
+            else :
+                jet.tags['L0'] = -99
+            if hasattr(event, 'S0') == True :
+                jet.tags['S0'] = event.S0
+            else :
+                jet.tags['S0'] = -99
+            if hasattr(event, 'Sp') == True :
+                jet.tags['Sp'] = event.Sp
+            else :
+                jet.tags['Sp'] = -99
+            if hasattr(event, 'Sm') == True :
+                jet.tags['Sm'] = event.Sm
+            else :
+                jet.tags['Sm'] = -99
+            if hasattr(event, 'Muons') == True :
+                jet.tags['Muons'] = event.Muons
+            else :
+                jet.tags['Muons'] = -99
+        
         
+    def beginLoop(self, setup):
+        super(ImpactParameter, self).beginLoop(setup)
+        if hasattr(self.cfg_ana, 'num_IP') == False :
+            self.tag_IP_b_LL = False
+        else :
+            if hasattr(self.cfg_ana, 'denom_IP') == False :
+                self.tag_IP_b_LL = False
+            else :
+                self.tag_IP_b_LL = True
+                self.num_IP_file = TFile.Open(self.cfg_ana.num_IP[0],"read")
+                self.num_IP_hist = self.num_IP_file.Get(self.cfg_ana.num_IP[1])
+                self.denom_IP_file = TFile.Open(self.cfg_ana.denom_IP[0],"read")
+                self.denom_IP_hist = self.denom_IP_file.Get(self.cfg_ana.denom_IP[1])
+                self.ratio_IP = TH1F("ratio_IP","num_IP over denom_IP", self.num_IP_hist.GetXaxis().GetNbins(), self.num_IP_hist.GetXaxis().GetXmin(), self.num_IP_hist.GetXaxis().GetXmax())
+                self.ratio_IP.Divide(self.num_IP_hist,self.denom_IP_hist)
+                #import pdb; pdb.set_trace()
+                
+                
+         
+        if hasattr(self.cfg_ana, 'num_IPs') == False :
```

Comment:
It seems you could generalize. 
I expect the LL tagger to work the same in all cases, whatever the distributions provided for num and denom. 
Therefore, there should be no need to duplicate code in this class. 

First, in this class, all duplicated code should be handled by specific utility methods, e.g. 

    load_LL_histograms(self.cfg_ana.num_IP)
    load_LL_histograms(self.cfg_ana.num_IPs)

And then in the same way when comes the time to compute the tagger. After this is done, we can discuss the possibility to support any number of LL taggers, but this is really second priority